musicbot/bot.py

```diff
@@ -7016,24 +7062,85 @@ async def cmd_debug(
             data = "\n".join(data.rstrip("`\n").split("\n")[1:])
 
         code = data.strip("` \n")
-
-        scope = globals().copy()
-        scope.update({"self": self})
-
         try:
-            result = eval(code, scope)  # pylint: disable=eval-used
+            run_type = "eval"
+            result = eval(code)  # pylint: disable=eval-used
+            log.debug("Debug code ran with eval().")
         except Exception:  # pylint: disable=broad-exception-caught
             try:
-                exec(code, scope)  # pylint: disable=exec-used
-            except Exception as e:  # pylint: disable=broad-exception-caught
-                traceback.print_exc(chain=False)
-                type_name = type(e).__name__
-                return Response(f"{type_name}: {str(e)}")
+                run_type = "exec"
+                # exec needs a fake locals so we can get `result` from it.
+                lscope: Dict[str, Any] = {}
+                # exec also needs locals() to be in globals() for access to work.
+                gscope = globals().copy()
+                gscope.update(locals().copy())
+                exec(code, gscope, lscope)  # pylint: disable=exec-used
+                log.debug("Debug code ran with exec().")
+                result = lscope.get("result", result)
+            except Exception as e:
+                log.exception("Debug code failed to execute.")
+                raise exceptions.CommandError(
+                    f"Failed to execute debug code.\n{codeblock.format(code)}\n"
+                    f"Exception: ```\n{type(e).__name__}:\n{str(e)}```"
+                ) from e
 
         if asyncio.iscoroutine(result):
             result = await result
 
-        return Response(codeblock.format(result))
+        return Response(f"**{run_type}() Result:**\n{codeblock.format(result)}")
+
+    @dev_only
+    async def cmd_makemarkdown(
+        self,
+        channel: MessageableChannel,
+        author: discord.Member,
+        cfg: str = "opts",
+    ) -> CommandResponse:
+        """
+        Command to generate markdown for options and permissions files.
+        Contents are generated from code and not pulled from the files!
+        """
+        valid_opts = ["opts", "perms"]
+        if cfg not in valid_opts:
```

Comment:
Is it bad practice to make this all one line? `if cfg not "opts" or "perms":`

Reply:
Not bad practice afaik.  The list here also gets used in generating the error message to show what the valid opts are.

Reply:
Oh sorry I missed that